drivers/dai/nxp/sai.c

```diff
@@ -0,0 +1,758 @@
+
+	device_map(&data->regmap, cfg->regmap_phys, cfg->regmap_size, K_MEM_CACHE_NONE);
+
+	/* enable clocks if any */
+	for (i = 0; i < cfg->clk_data.clock_num; i++) {
+		ret = clock_control_on(cfg->clk_data.dev,
+				       UINT_TO_POINTER(cfg->clk_data.clocks[i]));
+		if (ret < 0) {
+			return ret;
+		}
+
+		LOG_DBG("clock %s has been ungated", cfg->clk_data.clock_names[i]);
+	}
+
+	/* set TX/RX default states */
+	data->tx_state = DAI_STATE_NOT_READY;
+	data->rx_state = DAI_STATE_NOT_READY;
+
+	/* register ISR and enable IRQ */
+	cfg->irq_config();
+
+	return 0;
+}
+
+#define SAI_INIT(inst)								\
+										\
+BUILD_ASSERT(SAI_FIFO_DEPTH(inst) > 0 &&					\
+	     SAI_FIFO_DEPTH(inst) <= _SAI_FIFO_DEPTH(inst),			\
+	     "invalid FIFO depth");						\
+										\
+BUILD_ASSERT(SAI_RX_FIFO_WATERMARK(inst) > 0 &&					\
+	     SAI_RX_FIFO_WATERMARK(inst) <= _SAI_FIFO_DEPTH(inst),		\
+	     "invalid RX FIFO watermark");					\
+										\
+BUILD_ASSERT(SAI_TX_FIFO_WATERMARK(inst) > 0 &&					\
+	     SAI_TX_FIFO_WATERMARK(inst) <= _SAI_FIFO_DEPTH(inst),		\
+	     "invalid TX FIFO watermark");					\
+										\
+BUILD_ASSERT(IS_ENABLED(CONFIG_SAI_HAS_MCLK_CONFIG_OPTION) ||			\
+	     !DT_INST_PROP(inst, mclk_is_output),				\
+	     "SAI doesn't support MCLK config but mclk_is_output is specified");\
+										\
+static const struct dai_properties sai_tx_props_##inst = {			\
+	.fifo_address = SAI_TX_FIFO_BASE(inst),					\
+	.fifo_depth = SAI_FIFO_DEPTH(inst) * CONFIG_SAI_FIFO_WORD_SIZE,		\
+	.dma_hs_id = SAI_TX_DMA_MUX(inst),					\
+};										\
+										\
+static const struct dai_properties sai_rx_props_##inst = {			\
+	.fifo_address = SAI_RX_FIFO_BASE(inst),					\
+	.fifo_depth = SAI_FIFO_DEPTH(inst) * CONFIG_SAI_FIFO_WORD_SIZE,		\
+	.dma_hs_id = SAI_RX_DMA_MUX(inst),					\
+};										\
+										\
+void irq_config_##inst(void)							\
+{										\
+	IRQ_CONNECT(DT_INST_IRQN(inst),						\
+		    0,								\
+		    sai_isr,							\
+		    DEVICE_DT_INST_GET(inst),					\
+		    0);								\
+	irq_enable(DT_INST_IRQN(inst));						\
+}										\
+										\
+static struct sai_config sai_config_##inst = {					\
+	.regmap_phys = DT_INST_REG_ADDR(inst),					\
+	.regmap_size = DT_INST_REG_SIZE(inst),					\
+	.clk_data = SAI_CLOCK_DATA_DECLARE(inst),				\
+	.rx_fifo_watermark = SAI_RX_FIFO_WATERMARK(inst),			\
+	.tx_fifo_watermark = SAI_TX_FIFO_WATERMARK(inst),			\
+	.mclk_is_output = DT_INST_PROP_OR(inst, mclk_is_output, false),		\
+	.tx_props = &sai_tx_props_##inst,					\
+	.rx_props = &sai_rx_props_##inst,					\
+	.irq_config = irq_config_##inst,					\
+};										\
+										\
+static struct sai_data sai_data_##inst = {					\
+	.cfg.type = DAI_IMX_SAI,						\
+	.cfg.channels = 0,							\
+	.cfg.dai_index = DT_INST_PROP_OR(inst, dai_index, 0),			\
+};										\
+										\
+DEVICE_DT_INST_DEFINE(inst, &sai_init, NULL,					\
+		      &sai_data_##inst, &sai_config_##inst,			\
+		      PRE_KERNEL_1, CONFIG_DAI_INIT_PRIORITY,			\
```

Comment:
Is there a reason we want to change the stage at which this is initialized? Do you have a certain stage in mind?

Reply:
Do we need the SAI that early, in PRE_KERNEL_1?
Also, [here](https://github.com/zephyrproject-rtos/zephyr/pull/65187/files#diff-68ee1bfebdf2de1c63b56f306a143b0ea0a3e6501d0a88967406a171dbb9c014R388) you check for pre_kernel state.
I suggest POST_KERNEL.

Reply:
The stage check from `sai_tx_rx_disable()` will be eventually removed anyways once we split STOP into STOP and POST_STOP so we shouldn't necessarily decide based on that.

The main reason I went with `PRE_KERNEL_1` stage is because most drivers I've worked with are initialized at this stage. Also, this driver is meant to be generic so just because SOF allows a later initialization we shouldn't count on that for every use case. Despite this, I can't really give you an actual use case in which we'd prefer `PRE_KERNEL_1` instead of `POST_KERNEL` so if you still think this would be better then I wouldn't mind giving it a shot.

Reply:
Changed to `POST_KERNEL`